### PR TITLE
feat: Refactor DocumentationBuilder

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | 
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: features/custom_identifier.feature [hunk 1]
```diff
@@ -82,7 +82,7 @@ Feature: Using custom identifier on resource
       """
 
   Scenario: API doc is correctly generated
-    When I send a "GET" request to "/apidoc"
+    When I send a "GET" request to "/doc.jsonld"
```
Comment (dunglas) [inline features/custom_identifier.feature]: Can you keep `/doc` (or `/apidoc` if `/doc` is already taken by Swagger UI).

File: src/Documentation/Action/DocumentationAction.php [hunk 2]
```diff
@@ -11,27 +11,25 @@
...
- * @author Kévin Dunglas <dunglas@gmail.com>
+ * @author Amrouche Hamza <hamza.simperfit@gmail.com>
  */
 final class DocumentationAction
 {
-    private $apiDocumentationBuilder;
+    private $resourceNameCollectionFactory;
 
-    public function __construct(ApiDocumentationBuilderInterface $apiDocumentationBuilder)
+    public function __construct(ResourceNameCollectionFactoryInterface $resourceNameCollection)
```
Comment (dunglas) [inline src/Documentation/Action/DocumentationAction.php]: Should be `$resourceNameCollectionFactory`

File: src/Hydra/DocumentationNormalizer.php [hunk 3]
```diff
@@ -14,23 +14,27 @@
...
 /**
  * Creates a machine readable Hydra API documentation.
  *
  * @author Kévin Dunglas <dunglas@gmail.com>
  */
-final class ApiDocumentationBuilder implements ApiDocumentationBuilderInterface
+final class DocumentationNormalizer implements NormalizerInterface
 {
+    const FORMAT = 'jsonld';
+
```
Comment (dunglas) [inline src/Hydra/DocumentationNormalizer.php]: Extra blank line to remove

File: src/Hydra/DocumentationNormalizer.php [hunk 4]
```diff
@@ -57,12 +61,14 @@ public function __construct(ResourceNameCollectionFactoryInterface $resourceName
     /**
      * {@inheritdoc}
      */
-    public function getApiDocumentation() : array
+    public function normalize($object, $format = null, array $context = array())
     {
+        var_dump('toto');
```
Comment (dunglas) [inline src/Hydra/DocumentationNormalizer.php]: debug statement to remove

File: src/Hydra/DocumentationNormalizer.php [hunk 5]
```diff
@@ -236,7 +242,7 @@ public function getApiDocumentation() : array
             ],
         ];
 
-        $doc = ['@context' => $this->getContext(), '@id' => $this->urlGenerator->generate('api_hydra_doc')];
+        $doc = ['@context' => $this->getContext(), '@id' => $this->urlGenerator->generate('api_doc', ['_format' => 'jsonld'])];
```
Comment (dunglas) [inline src/Hydra/DocumentationNormalizer.php]: `'_format' => self::FORMAT`?

File: src/Hydra/DocumentationNormalizer.php [hunk 6]
```diff
@@ -404,7 +410,7 @@ private function getContext() : array
     {
         return
             [
-                '@vocab' => $this->urlGenerator->generate('api_hydra_doc', [], UrlGeneratorInterface::ABS_URL).'#',
+                '@vocab' => $this->urlGenerator->generate('api_doc', ['_format' => 'jsonld'], UrlGeneratorInterface::ABS_URL).'#',
```
Comment (dunglas) [inline src/Hydra/DocumentationNormalizer.php]: Same here (FORMAT const)

File: src/Hydra/DocumentationNormalizer.php [hunk 7]
```diff
@@ -417,4 +423,12 @@ private function getContext() : array
                 'returns' => ['@id' => 'hydra:returns', '@type' => '@id'],
             ];
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsNormalization($data, $format = null)
+    {
+        return self::FORMAT === $format && $data instanceof ResourceNameCollection;
```
Comment (dunglas) [inline src/Hydra/DocumentationNormalizer.php]: I'm afraid it will conflict with the EntrypointNormalizer.

File: src/JsonLd/ContextBuilder.php [hunk 8]
```diff
@@ -25,6 +25,7 @@
  */
 final class ContextBuilder implements ContextBuilderInterface
 {
+    const FORMAT = 'jsonld';
```
Comment (dunglas) [inline src/JsonLd/ContextBuilder.php]: Add an extra line after.

Comment (dunglas) [inline src/JsonLd/ContextBuilder.php]: Not sure the constant is needed here, however.

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php [hunk 9]
```diff
@@ -11,6 +11,7 @@
 
 namespace ApiPlatform\Core\Bridge\Symfony\Bundle\Command;
 
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceNameCollectionFactoryInterface;
 use ApiPlatform\Core\Swagger\ApiDocumentationBuilder;
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: Is now `DocumentationNormalizer`

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php [hunk 10]
```diff
@@ -24,13 +25,16 @@
 final class SwaggerCommand extends Command
 {
     private $apiDocumentationBuilder;
+    private $resourceNameCollectionFactory;
 
-    public function __construct(ApiDocumentationBuilder $apiDocumentationBuilder)
+
+    public function __construct(ApiDocumentationBuilder $apiDocumentationBuilder, ResourceNameCollectionFactoryInterface $resourceNameCollection)
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: Is now `DocumentationNormalizer`

**Comment by dunglas** [comment]
https://github.com/api-platform/core/blob/master/src/Bridge/Symfony/Bundle/Resources/views/SwaggerUi/index.html.twig#L30 must be updated to use the new route for the Swagger doc.

File: src/Bridge/Symfony/Bundle/Resources/config/routing/api.xml [hunk 11]
```diff
@@ -8,9 +8,17 @@
     <route id="api_entrypoint" path="/{index}.{_format}">
         <default key="_controller">api_platform.action.entrypoint</default>
         <default key="_format" />
+        <default key="_entrypoint">1</default>
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/routing/api.xml]: Now useless.

File: src/Bridge/Symfony/Bundle/Resources/config/routing/api.xml [hunk 12]
```diff
@@ -8,9 +8,17 @@
...
         <default key="_api_respond">1</default>
         <default key="index">index</default>
         <requirement key="index">index</requirement>
     </route>
 
+    <route id="api_doc" path="/apidoc.{_format}">
+        <default key="_controller">api_platform.action.documentation</default>
+        <default key="_api_respond">1</default>
+        <default key="_doc">1</default>
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/routing/api.xml]: Now useless too.

File: src/Bridge/Symfony/Bundle/Resources/config/swagger.xml [hunk 13]
```diff
@@ -18,19 +18,19 @@
             <argument>%api_platform.title%</argument>
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/swagger.xml]: Should be injected in the `DocumentationAction` and passed to the constructor of the new `Documentation` object.

File: src/Bridge/Symfony/Bundle/Resources/config/swagger.xml [hunk 14]
```diff
@@ -18,19 +18,19 @@
...
             <argument>%api_platform.description%</argument>
             <argument>%api_platform.version%</argument>
+
+            <tag name="serializer.normalizer" priority="32" />
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/swagger.xml]: Can be 16 isn't it?

File: src/Documentation/Documentation.php [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+use ApiPlatform\Core\Metadata\Resource\ResourceNameCollection;
+
+/**
+ * Generates the API documentation.
+ *
+ * @author Amrouche Hamza <hamza.simperfit@gmail.com>
+ */
+
+class Documentation
```
Comment (dunglas) [inline src/Documentation/Documentation.php]: `final`

File: src/Documentation/Documentation.php [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    public function __construct(ResourceNameCollection $resourceNameCollection)
+    {
+        $this->resourceNameCollection = $resourceNameCollection;
+    }
+
+   public function getResourceNameCollection() {
+       return $this->resourceNameCollection;
+   }
+
+}
```
Comment (dunglas) [inline src/Documentation/Documentation.php]: Add an extra blank line here.

File: src/Bridge/Symfony/Bundle/Resources/config/swagger.xml (hunk 13 above)
Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/Resources/config/swagger.xml]: done

**Comment by Simperfit** [comment]
ping @dunglas this is RFR

File: src/Action/EntrypointAction.php [hunk 17]
```diff
@@ -21,15 +21,17 @@
...
 final class EntrypointAction
 {
+    private $entrypoint;
     private $resourceNameCollectionFactory;
 
-    public function __construct(ResourceNameCollectionFactoryInterface $resourceNameCollection)
+    public function __construct(Entrypoint $entrypoint, ResourceNameCollectionFactoryInterface $resourceNameCollectionFactory)
     {
-        $this->resourceNameCollectionFactory = $resourceNameCollection;
+        $this->entrypoint = $entrypoint;
```
Comment (dunglas) [inline src/Action/EntrypointAction.php]: IMO the entrypoint should be a value object, not a service. Don't inject it, just return in `__invoke`: `new Entrypoint(...)`.

File: src/Api/Entrypoint.php [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * The first path you will see in the API.
+ *
+ * @author Amrouche Hamza <hamza.simperfit@gmail.com>
+ */
+final class Entrypoint
+{
+    private $resourceNameCollection;
+
+    public function create($resourceNameCollection): Entrypoint
```
Comment (dunglas) [inline src/Api/Entrypoint.php]: I would rename this method `__construct`. Missing typehint btw.

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php [hunk 19]
```diff
@@ -23,12 +25,16 @@
  */
 final class SwaggerCommand extends Command
 {
-    private $apiDocumentationBuilder;
+    private $documentationNormalizer;
+    private $resourceNameCollectionFactory;
+    private $documentation;
 
-    public function __construct(ApiDocumentationBuilder $apiDocumentationBuilder)
+    public function __construct(DocumentationNormalizer $documentationNormalizer, ResourceNameCollectionFactoryInterface $resourceNameCollection, Documentation $documentation)
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: Same here, `Documentation` should be a VO (with a `__construct` method).

File: src/Bridge/Symfony/Bundle/Resources/config/api.xml [hunk 20]
```diff
@@ -41,7 +41,13 @@
             <argument type="service" id="api_platform.router" />
             <argument type="service" id="api_platform.property_accessor" />
         </service>
-
+        <service id="api_platform.documentation" class="ApiPlatform\Core\Documentation\Documentation" public="false">
+            <argument>%api_platform.title%</argument>
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/api.xml]: So, those parameters should be injected in the action. The action will pass them to the VO constructor.

File: src/Bridge/Symfony/Bundle/Resources/views/SwaggerUi/index.html.twig [hunk 21]
```diff
@@ -27,7 +27,7 @@
     <script>
         $(function () {
             window.swaggerUi = new SwaggerUi({
-                url: '{{ path('api_swagger_doc') }}',
+                url: '{{ path('api_doc') }}',
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/views/SwaggerUi/index.html.twig]: IIUC, you should add {'_format': 'json'} as second argument (please test it).

File: src/Documentation/Action/DocumentationAction.php [hunk 22]
```diff
@@ -11,27 +11,27 @@
...
+ * @author Amrouche Hamza <hamza.simperfit@gmail.com>
  */
 final class DocumentationAction
 {
-    private $apiDocumentationBuilder;
+    private $documentation;
+    private $resourceNameCollectionFactory;
 
-    public function __construct(ApiDocumentationBuilderInterface $apiDocumentationBuilder)
+    public function __construct(Documentation $documentation, ResourceNameCollectionFactoryInterface $resourceNameCollectionFactory)
```
Comment (dunglas) [inline src/Documentation/Action/DocumentationAction.php]: Same here, Documentation isn't a service and shouldn't be injected.

File: src/Documentation/Documentation.php [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+        $this->description = $description;
+        $this->version = $version;
+        foreach ($formats as $format => $mimeTypes) {
+            foreach ($mimeTypes as $mimeType) {
+                $this->mimeTypes[] = $mimeType;
+            }
+        }
+    }
+
+    public function create($resourceNameCollection): Documentation
```
Comment (dunglas) [inline src/Documentation/Documentation.php]: Should be passed to the constructor as it is a VO.

File: src/Hal/Serializer/ResourceNameCollectionNormalizer.php [hunk 24]
```diff
@@ -11,11 +11,11 @@
```
Comment (dunglas) [inline src/Hal/Serializer/ResourceNameCollectionNormalizer.php]: Should be renamed `EntrypointNormalizer`.

**Comment by Simperfit** [comment]
ping @dunglas comments taken in account

**Comment by dunglas** [comment]
Thank you very much @Simperfit

**Pull request merged**